timeseries/src/main/scala/com/criteo/cuttle/timeseries/TimeSeriesScheduler.scala

```diff
@@ -287,66 +287,91 @@ case class TimeSeriesScheduler() extends Scheduler[TimeSeries] with TimeSeriesAp
       }
     }
 
-    def go(running: Set[Run]): Unit = {
-      val (completed, stillRunning) = running.partition(_._3.isCompleted)
+    def mainLoop(running: Set[Run]): Unit = {
+      val (completed, stillRunning) = running.partition {
+        case(_,_,effect) => effect.isCompleted
+      }
+
       val now = Instant.now
       val (stateSnapshot, completedBackfills, toRun) = atomic { implicit txn =>
-        completed.foreach {
-          case (job, context, future) =>
-            val jobState = if (future.value.get.isSuccess) Done else Todo(context.backfill)
-            _state() = _state() + (job ->
-              (_state().apply(job).update(context.toInterval, jobState)))
-        }
+        val (stateSnapshot, newBackfills, completedBackfills) =
+          computeNextState(workflow, now, _state(), _backfills(), completed)
```

Comment:
I think that name might be confusing because the state is still being updated using `jobsToRun`. Maybe something like `collectCompletedJobs` would be better?